functions for finding and using s2s from individual pmts

abotas opened the pull request:
These additions allow IC to find and interact with the s2s of individual pmts.

1.
`find_s12` has been broken up into 3 functions `find_peaks`, `select_peaks_of_allowed_length`, and `extract_peaks_from_waveform` (in python and cython). 
`find_s12` was too big. First it found peaks, then it selected peaks of the correct length, then it put those peaks into a dictionary. 
Without changing the inputs/outputs, adding flags, and other technical debt, `find_s12` would not allow us to efficiently find the s12s of individual pmts.
Note finding the s2s of individual pmts is not quite as simple as it first appears since we need to rebin each pmt `cwf` in the same exact way we rebinned the `csum`, so that the rebinned time bins of the s2pmtd match and sum to s2d for each peak.
Also, breaking `find_s12` into modular functions allows for much better testing. 

2. 
a class, `S2Pmt` has been added to `pmaps.pyx`. It is initialized with dictionaries `s2d` and `s2pmtd` which maps `peak_no -- > [[pmt0E], [pmt1E], ..., [pmtnE]]`. People should look at this and make sure they like the basic structure.

3. 
a function `get_s2pmtd` has been added to peak functions. It takes as input `CWF` and `peak_bounds`.  `peak_bounds`  is a dictionary mapping `peak_no --> i_start, i_stop` of the s2 peaks found in csum. Note `peak_bounds` is the output of `find_peaks`, and `get_s2pmtd` makes uses `extract_peaks_from_waveform` to construct s2pmtd --> tearing  `find_s12` apart was worth it. 

4. 
tests for all this stuff

`-----------`
I have only created the free functions, I have not yet started messing around with any of the cities or base cities, so I have not deleted find_s12. If people are happy i or whoever can update `pmap_city` and whatever else to use the new functions.

Commit:
add class S2Pmt

Commit:
write _find_peaks for more modular _find_s12

find_s12 is too big for no good reason. By splitting it up we are able
to easily find the s2pmt (s2s of the individual pmts). The code is
also more modular legible testable.

Commit:
add function extract_peaks_from_waveform  (last part of find_s12)

Commit:
test select_peaks_of_allowed_length

Commit:
test to make sure find_peaks finds the correct peak indx bounds

Commit:
test find_peaks does not find_peaks when index spaced by 1 + stride

Commit:
test extract_peaks_from_waveforms

Commit:
find_peaks does not have kwarg length

Commit:
cythonize find_peaks

Commit:
find_peaks DOES have kwarg length

Commit:
cythonize the rest of the modular find_s12 functions

Commit:
test cython functions

Commit:
write and test get_s2pmtd

jjgomezcadenas reviewed (commented):
Very good work. Some minor points:

1. find_s1 and find_s2 still use find_s12. I suggest making find_s12 private rather than delete it, then let find_s1 and find_s2 use the new functions and ensure in a test that results using old function and new functions is consistent.

2. in class S2Pmt I would change the function 

pmt_total_energy(self, int peak_number, int pmt_number)

 to

cpdef pmt_total_energy_in_peak(self, int peak_number, int pmt_number) 

and then add a new method:

cpdef pmt_total_energy(self, int pmt_number)

That adds the energy of all peaks.

3. Which of the three functions you have newly defined (if any) needs to be public (meaning leaving outside the python code)? My first reaction is that none of them, since they are all called internally by 
find_s1, find_s2 or get_s2pmtd, right? Make cdef of those functions that are not needed outside their ctyhon clients (unless you need to test them directly in a python code).

4. Have you tested that python and python code yield same results?

Anyway, again, excellent. Consider the above suggestions, rebase on top of the head and I will approve this very professional piece of work.

Commit:
imporove s2pmt methods

Commit:
make sure modular functions perform identically to find_s12

Commit:
add more tests for S2Pmt methods

Commit:
improved check that s2d and s2pmtd are consistent in S2Pmt

Commit:
add ic exception inconsistent s2d and s2pmtd

upon initialization of an S2Pmt, we assert that the energies in each
s2d peak are the sum of the energies across pmts in that peak in
s2pmtd, for each time bin.

jjgomezcadenas reviewed (commented):
Excellent. Are you still planning any addition? Otherwise I will approve.

Commit:
make find_peaks, select_peaks_..., and extract_peaks... cpdef

Commit:
have python find_s12 use modular functions

Commit:
accomodate rebin flag, for now

Commit:
make select_peaks_of_allowed_length private

jjgomezcadenas reviewed (approved):
Excellent work.
